src/utils/src/arg_parser.rs

```diff
@@ -0,0 +1,748 @@
+        let mut help_builder = vec![];
+
+        help_builder.push(format!("{} v{}.\n", self.name, self.version));
+
+        help_builder.push(format!("{}\n\n", self.header));
+
+        for arg in self.arguments.args.values() {
+            help_builder.push(format!("{}\n", arg.format_help()));
+        }
+
+        help_builder.concat()
+    }
+
+    /// Return a reference to `arguments` field.
+    pub fn arguments(&self) -> &Arguments {
+        &self.arguments
+    }
+}
+
+/// Stores the characteristics of the `name` command line argument.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ArgInfo<'a> {
+    name: &'a str,
+    required: bool,
+    requires: Option<&'a str>,
+    takes_value: bool,
+    default_value: Option<Value>,
+    help: Option<&'a str>,
+    user_value: Option<Value>,
+}
+
+impl<'a> ArgInfo<'a> {
+    /// Create a new `ArgInfo` that keeps the necessary information for an argument.
+    pub fn new(name: &'a str) -> ArgInfo<'a> {
+        ArgInfo {
+            name,
+            required: false,
+            requires: None,
+            takes_value: false,
+            default_value: None,
+            help: None,
+            user_value: None,
+        }
+    }
+
+    /// Set if the argument *must* be provided by user.
+    pub fn required(mut self, required: bool) -> Self {
+        self.required = required;
+        self
+    }
+
+    /// Add `other_arg` as a required parameter when `self` is specified.
+    pub fn requires(mut self, other_arg: &'a str) -> Self {
+        self.requires = Some(other_arg);
+        self
+    }
+
+    /// If `takes_value` is true, then the user *must* provide a value for the
+    /// argument, otherwise that argument is a flag.
+    pub fn takes_value(mut self, takes_value: bool) -> Self {
+        self.takes_value = takes_value;
+        self
+    }
+
+    /// Keep a default value which will be used if the user didn't provide a value for
+    /// the argument.
+    pub fn default_value(mut self, default_value: &'a str) -> Self {
+        self.default_value = Some(Value::String(String::from(default_value)));
+        self
+    }
+
+    /// Set the information that will be displayed for the argument when user passes
+    /// `--help` flag.
+    pub fn help(mut self, help: &'a str) -> Self {
+        self.help = Some(help);
+        self
+    }
+
+    fn format_help(&self) -> String {
+        let mut help_builder = vec![];
+
+        help_builder.push(format!("--{}", self.name));
+
+        if self.takes_value {
+            help_builder.push(format!(" <{}>", self.name));
+        }
+
+        if let Some(help) = self.help {
+            help_builder.push(format!(": {}", help));
+        }
+        help_builder.concat()
+    }
+}
+
+/// Represents the value of an argument, which will be a `String` if
+/// the argument takes a value, or `bool` if it's a flag.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Value {
+    Bool(bool),
+    String(String),
+}
```

Comment:
On a second thought a better suggestion would be to use something like this:

```
pub struct ArgInfo<'a> {
    name: &'a str,
    required: bool,
    requires: Option<&'a str>,
    help: Option<&'a str>,

    value_info: ValueInfo,
}

pub enum ValueInfo {
    Flag(FlagValueInfo),
    Input(InputValueInfo),
}

pub struct FlagValueInfo {
    default_value: bool,
    user_value: bool,
}

pub struct InputValueInfo {
    default_value: Option<String>,
    user_value: Option<String>,
}
```

This would eliminate the need for the `takes_value`, and would ensure that `ArgInfo` doesn't contain invalid data. For example not you can have an `ArgInfo` instance with the following data:

```
ArgInfo<'a> {
    ...
    takes_value: false,
    default_value: "value",
    user_value: "x",
}
```

or

```
ArgInfo<'a> {
    ...
    takes_value: true,
    default_value: true,
    user_value: "x",
}
```

But it might be a lot of work to implement and it might not be worth it. So feel free to discard the suggestion.

Reply:
I find it more clear as it is right now. We duplicate the information if both boolean and values that take values have separate structures with the same fields.

I would propose to iterate over this in a next PR because there are still a few tiny things we need to fix anyway.

Reply:
Ok. Sounds good to me.